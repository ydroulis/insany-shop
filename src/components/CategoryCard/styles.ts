"use client"

import Link from 'next/link'
import styled from 'styled-components'

export const CardLink = styled(Link)`
    display: flex;
    height: 100%;
    min-height: 116px;
    text-decoration: none;
    color: inherit;

    border: 1px solid #E0E0E0;

    width: 150px;
    padding: 0rem 1rem;

    border-radius: 12px;
    background-color: #fff;
    transition: transform 0.3s ease-in-out;

    color: #000000;
    font-size: 1rem;

    &:hover {
        transform: scale(1.02);
    }

    &:focus-visible {
        outline: 3px solid #2563eb;
        outline-offset: 4px;
        border-radius: 12px;
        transform: scale(1.02);
    }

    @media (min-width: 768px) {
        width: 220px;
        padding: 0rem 1rem;
    }

    @media (min-width: 1024px) {
        width: 177px;
        padding: 0rem 1rem;
    }

    @media (min-width: 1440px) {
        width: 208px;
    }
`

export const CardArticle = styled.article`
    width: 100%;
    height: 100%;

    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    gap: 16px;
`

export const Title = styled.h3`
    font-weight: 700;
    text-align: center;
`

export const ProductCount = styled.p`
    text-align: center;
`

