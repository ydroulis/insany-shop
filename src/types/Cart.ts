
export type CartProduct = {
    id: number
    name: string
    description: string
    price: number
    image: string
    stock: number
    items?: number
}

export type Cart = {
    content: CartProduct[]
    total: number
}