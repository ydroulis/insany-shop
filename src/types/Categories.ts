export type Categories = Category[];

export type Category = {
    id: string,
    name: string,
    description: string,
    icon: string,
    productCount: number
};