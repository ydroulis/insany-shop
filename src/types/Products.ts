export type Products = Product[]

export type Product = {
    id: number,
    name: string,
    description: string,
    price: number,
    image: string,
    category: string,
    stock: number,
    rating: number,
    brand?: string

}