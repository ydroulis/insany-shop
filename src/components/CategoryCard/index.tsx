import { Category } from '@/types/Categories';
import React from 'react';

import * as S from './styles';

const CategoryCard: React.FC<Category> = ({ id, productCount, name }) => {
    const productLabel = `${productCount} ${productCount > 1 ? 'produtos' : 'produto'}`;

    return (
        <S.CardLink
            href={`/category/${id}`}
            aria-label={`Ver categoria ${name} com ${productLabel}`}
        >
            <S.CardArticle data-testid="category-card">
                <S.Title>{name}</S.Title>
                <S.ProductCount aria-hidden="true">{productLabel}</S.ProductCount>
            </S.CardArticle>
        </S.CardLink>
    );
}

export default CategoryCard;