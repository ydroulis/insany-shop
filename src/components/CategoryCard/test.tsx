import React from "react";
import { render, screen } from "@testing-library/react";
import CategoryCard from "./";
import { Category } from "@/types/Categories";
import "@testing-library/jest-dom";

// Mock do next/link (necessário em testes com Next.js)
jest.mock("next/link", () => {
    const Link = ({ children, href, ...rest }: { children: React.ReactNode; href: string; }) => (
        <a href={href} {...rest}>
            {children}
        </a>
    );
    return Link;
});

describe("<CategoryCard />", () => {
    const mockCategory: Category = {
        id: "eletronicos",
        name: "Eletrônicos",
        productCount: 5,
        description: "",
        icon: ""
    };

    it("should render category name", () => {
        render(<CategoryCard {...mockCategory} />);
        expect(
            screen.getByRole("heading", { name: /eletrônicos/i })
        ).toBeInTheDocument();
    });

    it("should render product count in plural", () => {
        render(<CategoryCard {...mockCategory} />);
        expect(screen.getByText("5 produtos")).toBeInTheDocument();
    });

    it("should render product count in singular ", () => {
        render(<CategoryCard {...mockCategory} productCount={1} />);
        expect(screen.getByText("1 produto")).toBeInTheDocument();
    });

    it("should contain an accessible link properly", () => {
        render(<CategoryCard {...mockCategory} />);
        const link = screen.getByRole("link", {
            name: "Ver categoria Eletrônicos com 5 produtos",
        });
        expect(link).toBeInTheDocument();
        expect(link).toHaveAttribute("href", "/category/eletronicos");
    });

    it("should match snapshot", () => {
        const { container } = render(<CategoryCard {...mockCategory} />);
        expect(container).toMatchSnapshot();
    });
});
