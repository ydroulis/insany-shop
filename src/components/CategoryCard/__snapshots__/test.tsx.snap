// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<CategoryCard /> should match snapshot 1`] = `
<div>
  <a
    aria-label="Ver categoria Eletrônicos com 5 produtos"
    class="sc-bRKDuR hqlWNC"
    href="/category/eletronicos"
  >
    <article
      class="sc-hvigdm dBOyIL"
      data-testid="category-card"
    >
      <h3
        class="sc-fhHczv kQgZIE"
      >
        Eletrônicos
      </h3>
      <p
        aria-hidden="true"
        class="sc-ggWZvA hvwsfd"
      >
        5 produtos
      </p>
    </article>
  </a>
</div>
`;
